File: Haskell/euler-698-123-numbers.hs
import Data.Char (digitToInt)
import Data.List (findIndices)
import System.Environment (getArgs)
import Text.Read (readMaybe) 

-- Check if a number is a valid 123-number
is123Number :: Int -> Bool
is123Number 1 = True
is123Number 2 = True
is123Number 3 = True
is123Number n
  | n > 3 = let digits = map digitToInt $ show n
                validDigits = all (`elem` [1, 2, 3]) digits
                counts = map (\d -> length $ findIndices (== d) digits) digits
            in validDigits && all is123Number counts
  | otherwise = False

-- Generate the sequence of 123-numbers up to the N-th number
find123Number :: Int -> Maybe Int
find123Number n = 
  let upperLimit = 1000  -- Arbitrary upper limit for brute force approach
      sequence123 = filter is123Number [1..upperLimit]
  in if length sequence123 >= n then Just (sequence123 !! (n - 1)) else Nothing

-- Modulo operation
modulo :: Int -> Int
modulo number = number `mod` 123123123

-- Main function
main :: IO ()
main = do
  args <- getArgs
  case args of
    [arg] -> case readMaybe arg of
               Just n -> case find123Number n of
                           Just number -> print $ modulo number
                           Nothing -> putStrLn "Error: Unable to find the 123-number."
               Nothing -> putStrLn "Error: Invalid input. Please enter a valid number."
    _ -> putStrLn "Usage: Euler698 <n>"

File: Perl/euler-698-123-numbers.pl
% Check if a number is a valid 123-number
is_123_number(1).
is_123_number(2).
is_123_number(3).
is_123_number(N) :-
    N > 3,
    number_chars(N, Digits),
    all_members(Digits, ['1', '2', '3']),
    findall(X, (member(D, Digits), number_chars(X, [D])), Counts),
    all_123_numbers(Counts).

% Check if all elements in the list are 123-numbers
all_123_numbers([]).
all_123_numbers([H|T]) :-
    is_123_number(H),
    all_123_numbers(T).

% Helper predicate to check if all elements of the first list are members of the second list
all_members([], _).
all_members([H|T], List) :-
    member(H, List),
    all_members(T, List).

% Find the N-th 123-number
find_123_number(N, Number) :-
    UpperLimit is 1000,  % Arbitrary upper limit for brute force approach
    numlist(1, UpperLimit, Range),
    include(is_123_number, Range, Seq),
    nth1(N, Seq, Number).

% Modulo operation
modulo(Number, ModuloResult) :-
    ModuloBase = 123123123,
    ModuloResult is Number mod ModuloBase.

% Main entry point
% Main entry point
main :-
    current_prolog_flag(argv, Arguments),
    (   Arguments = [Arg],
        atom_number(Arg, N) ->
        write('Finding the '), write(N), write('th 123-number...'), nl,
        (   find_123_number(N, Number) ->
            write('Found number: '), write(Number), nl,
            modulo(Number, ModuloResult),
            write('Modulo result: '), write(ModuloResult), nl
        ;   write('Error: Unable to find the 123-number.'), nl
        )
    ;   write('Error: Invalid command line arguments. Please provide a single numeric argument.'), nl
    ).


% Direct the interpreter to execute the main predicate upon startup.
:- initialization(main).
